src/completions.jl

```diff
@@ -1,74 +1,102 @@
+### baseline completions ###
+
 handle("completions") do data
-  @destruct [path || nothing,
-             mod || "Main",
-             editorContent || "",
-             lineNumber || 1,
-             startLine || 0,
-             column || 1,
-             line, force] = data
+  @destruct [
+    # general
+    line,
+    mod || "Main",
+    path || nothing,
+    # local context
+    context || "",
+    row || 1,
+    startRow || 0,
+    column || 1,
+    # configurations
+    force || false
+  ] = data
 
   withpath(path) do
-    m = getmodule(mod)
-
-    cs, pre = basecompletionadapter(line, m, force, lineNumber - startLine, column, editorContent)
-
-    Dict(:completions => cs,
-         :prefix      => string(pre))
+    basecompletionadapter(
+      # general
+      line, mod,
+      # local context
+      context, row - startRow, column,
+      # configurations
+      force
+    )
   end
 end
 
 using REPL.REPLCompletions
 
-function basecompletionadapter(line, mod, force, lineNumber, column, text)
-  comps, replace, shouldcomplete = try
+const CompletionSuggetion = Dict{Symbol, String}
+
+# as an heuristic, suppress completions if there are over 500 completions,
+# ref: currently `completions("", 0)` returns **1132** completions as of v1.3
+const SUPPRESS_COMPLETION_THRESHOLD = 500
+
+# autocomplete-plus only shows at most 200 completions
+# ref: https://github.com/atom/autocomplete-plus/blob/master/lib/suggestion-list-element.js#L49
+const MAX_COMPLETIONS = 200
+
+function basecompletionadapter(
+  # general
+  line, m = "Main",
+  # local context
+  context = "", row = 1, column = 1,
+  # configurations
+  force = false
+)
+  mod = getmodule(m)
+
+  cs, replace, shouldcomplete = try
     completions(line, lastindex(line), mod)
   catch err
     # might error when e.g. type inference fails
-    [], 1:0, false
+    REPLCompletions.Completion[], 1:0, false
   end
 
-  # Suppress completions if there are too many of them unless activated manually
-  # @TODO: Checking whether `line` is a valid text to be completed in atom-julia-client
-  #        in advance and drop this check
-  if !force && length(comps) > MAX_COMPLETIONS
-    comps = []
+  # suppress completions if there are too many of them unless activated manually
+  # e.g. when invoked with `$|`, `(|`, etc.
+  # TODO: check whether `line` is a valid text to complete in frontend
+  if !force && length(cs) > SUPPRESS_COMPLETION_THRESHOLD
+    cs = REPLCompletions.Completion[]
     replace = 1:0
   end
 
-  pre = line[replace]
-  d = []
-  for c in comps
-    if REPLCompletions.afterusing(line, Int(first(replace))) # need `Int` for correct dispatch on x86
-      c isa REPLCompletions.PackageCompletion || continue
-    end
-    try
-      push!(d, completion(mod, c))
-    catch err
-      continue
-    end
```

Comment:
I removed this exception check since now our baseline completions are (hopefully) error safe.